src/elf/parser.rs

```diff
@@ -341,17 +342,30 @@ pub(crate) struct ElfParser {
     ///         to make sure we never end up with a dangling reference.
     cache: RefCell<Cache<'static>>,
     /// The memory mapped file.
-    _mmap: Mmap,
+    _mmap: Rc<Mmap>,
 }
 
 impl ElfParser {
+    /// Create an `ElfParser` from an open file.
     pub fn open_file(file: File) -> Result<ElfParser, Error> {
-        let mmap = Mmap::map(&file)?;
+        let mmap = Rc::new(Mmap::map(&file)?);
+        let offset = 0;
+        Self::from_mmap(mmap, offset)
+    }
+
+    /// Create an `ElfParser` from mmap'ed data.
+    pub fn from_mmap(mmap: Rc<Mmap>, offset: usize) -> Result<ElfParser, Error> {
```

Comment:
so the idea here is to take existing mmap and work with a subset of it, starting from offset, right? But why we do not limit the end as well? In general, embedded ELF file within APK won't go all the way to the end of mmap. Not specifying size/end_offset seems very dangerous and error-prone.

Reply:
> so the idea here is to take existing mmap and work with a subset of it, starting from offset, right?

Correct.

> But why we do not limit the end as well? In general, embedded ELF file within APK won't go all the way to the end of mmap. Not specifying size/end_offset seems very dangerous and error-prone.

ELF carries its own size information. So it would be duplicated data without real value. Not sure I follow what is dangerous or error prone about it. Can you elaborate, please?

Reply:
In my mind it's like taking a slice from an array. You always specify [from .. to), so it feels natural to have the same interface her. It's error-prone if ELF itself is corrupted or the processing code is buggy, we might not notice that we went out of bounds of intended piece of mmap. Just seems prudent to specify what memory range we are intending to work with.

Unless it's problematic to know the size upfront? I assumed we do know from ZIP file itself?

Reply:
I don't think it's problematic, just more cumbersome for no apparent benefit. In general I agree that we work with slices that are restricted properly on both ends. But given that an ELF file is already a container if you will, restricting the end of the range doesn't add much. Yes, the file can be corrupted, but still nothing bad would happen. Worst case we read outside if the ELF file itself and report bogus data for an already bogus file ¯\_(°ペ)_/¯
We use this function from `ElfParser::open_file` and there we don't have the size trivially available (we could get it, but there it's *truly* a waste, because the mmap already covers only the file and nothing else). Ideally we'd let callers pass in a [`RangeBounds`](https://doc.rust-lang.org/std/ops/trait.RangeBounds.html), but unfortunately we can't handle that in there properly.